util/src/main/java/tech/pegasys/artemis/util/async/SafeFuture.java

```diff
@@ -0,0 +1,160 @@
+    future.complete(value);
+    return future;
+  }
+
+  public static <U> SafeFuture<U> failedFuture(Throwable ex) {
+    SafeFuture<U> future = new SafeFuture<>();
+    future.completeExceptionally(ex);
+    return future;
+  }
+
+  public static <U> SafeFuture<U> of(final CompletionStage<U> stage) {
+    if (stage instanceof SafeFuture) {
+      return (SafeFuture<U>) stage;
+    }
+    final SafeFuture<U> safeFuture = new SafeFuture<>();
+    stage.whenComplete(
+        (result, error) -> {
+          if (error != null) {
+            safeFuture.completeExceptionally(error);
+          } else {
+            safeFuture.complete(result);
+          }
+        });
+    return safeFuture;
+  }
+
+  public static <U> SafeFuture<Void> allOf(final SafeFuture<?>... futures) {
+    return of(CompletableFuture.allOf(futures));
+  }
+
+  public static SafeFuture<Object> anyOf(final SafeFuture<?>... futures) {
+    return of(CompletableFuture.anyOf(futures));
+  }
+
+  @Override
+  public <U> SafeFuture<U> newIncompleteFuture() {
+    return new SafeFuture<>();
+  }
+
+  public void reportExceptions() {
+    reportExceptions(this);
+  }
+
+  public void finish(final Runnable onSuccess) {
+    finish(complete -> onSuccess.run());
+  }
+
+  public void finish(final Consumer<T> onSuccess) {
+    reportExceptions(thenAccept(onSuccess));
+  }
+
+  public void finish(final Runnable onSuccess, final Consumer<Throwable> onError) {
+    finish(result -> onSuccess.run(), onError);
+  }
+
+  public void finish(final Consumer<T> onSuccess, final Consumer<Throwable> onError) {
+    whenComplete(
+            (result, error) -> {
+              if (error != null) {
+                onError.accept(error);
+              } else {
+                onSuccess.accept(result);
+              }
+            })
+        .reportExceptions();
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <U> SafeFuture<U> thenApply(final Function<? super T, ? extends U> fn) {
+    return (SafeFuture<U>) super.thenApply(fn);
+  }
+
+  @Override
+  public SafeFuture<Void> thenRun(final Runnable action) {
+    return (SafeFuture<Void>) super.thenRun(action);
+  }
+
+  @Override
+  public SafeFuture<Void> thenAccept(final Consumer<? super T> action) {
+    return (SafeFuture<Void>) super.thenAccept(action);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <U, V> SafeFuture<V> thenCombine(
+      final CompletionStage<? extends U> other,
+      final BiFunction<? super T, ? super U, ? extends V> fn) {
+    return (SafeFuture<V>) super.thenCombine(other, fn);
+  }
+
+  @Override
+  public <U> SafeFuture<U> thenCompose(final Function<? super T, ? extends CompletionStage<U>> fn) {
+    return (SafeFuture<U>) super.thenCompose(fn);
+  }
+
+  @Override
+  public SafeFuture<T> exceptionally(final Function<Throwable, ? extends T> fn) {
+    return (SafeFuture<T>) super.exceptionally(fn);
+  }
+
+  @SuppressWarnings("unchecked")
```

Comment:
Should we add some tests where the warnings are suppressed - to make sure the type comes out as expected?

Reply:
Probably worth one as a sanity check - `CompletableFuture` does guarantee that it will always use `newIncompleteFuture` to create new instances specifically so you can do this kind of thing. Tests all over the place will fail if it stops working.